Source/Dafny/Compilers/Compiler-python.cs

```diff
@@ -205,7 +205,43 @@ public class PythonCompiler : SinglePassCompiler {
     }
 
     protected override ConcreteSyntaxTree CreateIterator(IteratorDecl iter, ConcreteSyntaxTree wr) {
-      throw new UnsupportedFeatureException(Token.NoToken, Feature.Iterators);
+      var cw = CreateClass(IdProtect(iter.EnclosingModuleDefinition.CompileName), IdName(iter), false,
+        IdName(iter), iter.TypeArgs, iter, null, iter.tok, wr) as ClassWriter;
+      Debug.Assert(cw != null, nameof(cw) + " != null");
+      var constructorWriter = cw.ConstructorWriter;
+      var w = cw.MethodWriter;
+      // here come the fields
+      Constructor ct = null;
+      foreach (var member in iter.Members) {
+        switch (member) {
+          case Field { IsGhost: false } f:
+            DeclareField(IdName(f), false, false, f.Type, f.tok, PlaceboValue(f.Type, constructorWriter, f.tok, true), constructorWriter);
+            break;
+          case Constructor constructor:
+            Contract.Assert(ct == null);  // we're expecting just one constructor
+            ct = constructor;
+            break;
+        }
+      }
+      Contract.Assert(ct != null);  // we do expect a constructor
+      constructorWriter.WriteLine("self._iter = None");
+
+      var nonNullIns = ct.Ins.Where(f => !f.IsGhost).ToList();
+      var args = nonNullIns.Select(IdName).Prepend("self").Comma();
+      var wCtor = w.NewBlockPy($"def {IdName(ct)}({args}):");
+      foreach (var p in nonNullIns) {
+        wCtor.WriteLine("self.{0} = {0}", IdName(p));
+      }
+      wCtor.WriteLine("self._iter = self.TheIterator()");
+
+      var wMoveNext = w.NewBlockPy("def MoveNext(self):");
+      wMoveNext.NewBlockPy("try:")
+        .WriteLine("next(self._iter)")
+        .WriteLine("return True");
+      wMoveNext.NewBlockPy("except (StopIteration, TypeError) as e:")
+        .WriteLine("return False");
```

Comment:
Could this be put in a base class that iterators inherit from?

Reply:
Yes, but it's not quite as easy as I would like it to be. The `CreateClass` wants the superclasses as `Type`s. There's also the argument to be made the the expected number of iterators per program is likely to be <1.